PWGHF/D2H/Tasks/taskDirectedFlowCharmHadrons.cxx

```diff
@@ -0,0 +1,304 @@
+        cent = collision.centFV0A();
+        break;
+    }
+    return cent;
+  }
+
+  /// Check if the collision is selected
+  /// \param collision is the collision with the Q vector information
+  /// \param bc is the bunch crossing with timestamp information
+  /// \return true if the collision is selected, false otherwise
+  template <o2::hf_centrality::CentralityEstimator centEstimator>
+  bool isCollSelected(CollsWithQvecs::iterator const& collision,
+                      aod::BCsWithTimestamps const&)
+  {
+    double centrality{-1.f};
+    const auto rejectionMask = hfEvSel.getHfCollisionRejectionMask<true, centEstimator, aod::BCsWithTimestamps>(collision, centrality, ccdb, registry);
+
+    /// monitor the satisfied event selections
+    hfEvSel.fillHistograms(collision, rejectionMask, centrality);
+    return rejectionMask == 0;
+  }
+
+  /// Get the Q vector
+  /// \param collision is the collision with the Q vector information
+  std::vector<double> getQvec(CollsWithQvecs::iterator const& collision)
+  {
+    double qxAQVec = 0.0;
+    double qyAQVec = 0.0;
+    double qxCQVec = 0.0;
+    double qyCQVec = 0.0;
+    qxAQVec = collision.qxZDCA();
+    qyAQVec = collision.qyZDCA();
+    qxCQVec = collision.qxZDCC();
+    qyCQVec = collision.qyZDCC();
+    return {qxAQVec, qyAQVec, qxCQVec, qyCQVec};
+  }
+
+  /// Compute the scalar product
+  /// \param collision is the collision with the Q vector information and event plane
+  /// \param candidates are the selected candidates
+  template <DecayChannel channel, typename T1, typename Trk>
+  void runFlowAnalysis(CollsWithQvecs::iterator const& collision,
+                       T1 const& candidates, Trk const& /*tracks*/)
+  {
+    double cent = getCentrality(collision);
+    if (cent < centralityMin || cent > centralityMax) {
+      return;
+    }
+
+    if (!collision.triggerevent()) { // for selecting only callibrated events
+      return;
+    }
+
+    std::vector<double> qVecs = getQvec(collision);
+    double qxAQVec = qVecs[0];
+    double qyAQVec = qVecs[1];
+    double qxCQVec = qVecs[2];
+    double qyCQVec = qVecs[3];
+
+    auto qxZDCA = qxAQVec;
+    auto qyZDCA = qyAQVec;
+    auto qxZDCC = qxCQVec;
+    auto qyZDCC = qyCQVec;
+
+    auto QxtQxp = qxZDCC * qxZDCA;
+    auto QytQyp = qyZDCC * qyZDCA;
+    auto QxpQyt = qxZDCA * qyZDCC;
+    auto QxtQyp = qxZDCC * qyZDCA;
+
+    // denominators for SP calculation
+    registry.fill(HIST("hpQxtQxpvscent"), cent, QxtQxp);
+    registry.fill(HIST("hpQytQypvscent"), cent, QytQyp);
+    registry.fill(HIST("hpQxpQytvscent"), cent, QxpQyt);
+    registry.fill(HIST("hpQxtQypvscent"), cent, QxtQyp);
+
+    for (const auto& candidate : candidates) {
+      double massCand = 0.;
+      std::vector<double> outputMl = {-999., -999.};
+      if constexpr (std::is_same_v<T1, CandDplusData> || std::is_same_v<T1, CandDplusDataWMl>) {
+        massCand = hfHelper.invMassDplusToPiKPi(candidate);
```

Comment:
Do you need this because you want to support multiple species?

Reply:
[Reply]: Yes for that reason I would like to keep it.